java-client/barrage/src/main/java/io/deephaven/client/impl/BarrageSubscriptionImpl.java

```diff
@@ -454,32 +466,153 @@ private interface FutureAdapter extends Future<Table> {
         boolean complete(Table value);
 
         boolean completeExceptionally(Throwable ex);
+
+        /**
+         * Called when the hand-off from the future is complete to release the construction scope.
+         */
+        void maybeRelease();
+
+        @FunctionalInterface
+        interface Supplier {
+            Table get() throws InterruptedException, ExecutionException, TimeoutException;
+        }
+
+        @FinalDefault
+        default Table doGet(final Supplier supplier) throws InterruptedException, ExecutionException, TimeoutException {
+            try {
+                final Table result = supplier.get();
+
+                if (result instanceof LivenessArtifact) {
+                    ((LivenessArtifact) result).manageWithCurrentScope();
+                }
+
+                return result;
+            } finally {
+                maybeRelease();
+            }
```

Comment:
I think we should not be releasing in the case of a timeout. In that scenario, it remains the future's owner's responsibility to cancel or get the future.